File: webgl/source/js/renders/bumpingsphererender.js
class BumpingSphereRender extends Render {
    withBump = true;
    matrices = new Matrices();
    prog
    progWithBump
    sphere = new GlObject();

    rotY = 1.5 * Math.PI / 180;
    rotX = Math.PI / 180;

    light = new LightDirectional(glMatrix.vec3.fromValues(0.3, 0.3, 0.3),
        glMatrix.vec3.fromValues(1, 1, 1),
        glMatrix.vec3.fromValues(0.85, 0.8, 0.75));

    onSetup(appOGL) {

        glMatrix.vec3.normalize(this.light.direction, this.light.direction);

        const gl = appOGL.gl;

        this.prog = new ShaderProgramBuilder(gl)
            .lightDirectional(false)
            .texture2D()
            .build();

        this.progWithBump = new ShaderProgramBuilder(gl)
            .lightDirectional(true)
            .texture2D()
            .build();

        gl.useProgram(this.prog.idProgram);

        this.sphere.model = Models
            .sphere(gl, 1,
                36, 36, 1, 0.5, 0.5, "sphere")
            .calcTangent()
            .toGlModel(gl);

        //this.sphere.texture = GlTexture.newTexture2D(gl, "images/textures/235.jpg", "test");
        this.sphere.texture = GlTexture.newTexture2D(gl, "images/textures/bricks/bricks053_1k.jpg", "bricks");
        this.sphere.normalTexture = GlTexture.newTexture2D(gl, "images/textures/bricks/bricks053_1k_normal.jpg", "bricks-normal");

        this.sphere.transforms = glMatrix.mat4.create();
        glMatrix.mat4.translate(this.sphere.transforms, this.sphere.transforms, [0, 0, -6]);

        gl.clearColor(0.5, 0.5, 0.5, 1.0); // Clear to black, fully opaque
        gl.clearDepth(1.0); // Clear everything
        gl.enable(gl.DEPTH_TEST); // Enable depth testing
        gl.depthFunc(gl.LEQUAL); // Near things obscure far things
    }

    onDrawFrame(appOGL) {
        const gl = appOGL.gl;
        gl.clear(gl.COLOR_BUFFER_BIT | gl.DEPTH_BUFFER_BIT);

        const prog = this.withBump ? this.progWithBump : this.prog;

        glMatrix.mat4.rotate(this.sphere.transforms, this.sphere.transforms,
            this.rotX, [1, 0, 0]);
        glMatrix.mat4.rotate(this.sphere.transforms, this.sphere.transforms,
            this.rotY, [0, 1, 0]);

        gl.useProgram(prog.idProgram);
        prog.uniform(this.sphere, this.matrices, this.light);
        this.sphere.model.draw(gl);
    }

}

File: android/app/src/main/java/com/darkraha/gldemos/renders/BumpingSphereRender.kt
package com.darkraha.gldemos.renders

import com.darkraha.opengldemokt.gl.shader.ShaderProgramBuilder
import org.joml.Matrix4f
import org.joml.Vector3f
import javax.microedition.khronos.opengles.GL10
import android.opengl.GLES30.*
import com.darkraha.gldemos.R
import com.darkraha.gldemos.gl.*
import com.darkraha.gldemos.gl.GlCommon.TO_RAD
import com.darkraha.gldemos.gl.modelling.Models
import javax.microedition.khronos.egl.EGLConfig

class BumpingSphereRender : Render() {
    private val rotY = 1.5f * TO_RAD
    private val rotX = TO_RAD
    private val matrices: Matrices = Matrices()
    private lateinit var prog: ShaderProgram
    private val sphere: GlObject = GlObject()

    private val light: LightDirectional = LightDirectional(
        Vector3f(0.6f, 0.6f, 0.6f),
        Vector3f(1.0f, 1.0f, 1.0f),
        Vector3f(0.85f, 0.8f, 0.75f).normalize()
    )

    override fun onSurfaceChanged(gl: GL10?, width: Int, height: Int) {
        super.onSurfaceChanged(gl, width, height)
        matrices.projection.identity().perspective(GlCommon.ALNGLE45, aspect, 1f, 100f)
    }

    override fun onSurfaceCreated(gl: GL10?, config: EGLConfig?) {
        prog = ShaderProgramBuilder()
            .texture2D()
            .lightDirectional(true)
            .matrix()
            .build()

        sphere.model = Models
            .sphere(
                1f,
                36, 36,
                1f, 0.5f, 0.5f,
                "sphere-0"
            )
            .calcTangent()
            .toGlModel()

        sphere.transforms = Matrix4f()
        sphere.transforms!!.translate(0f, 0f, -6f)
        sphere.texture = GlTexture.newTexture2D(
            R.drawable.bricks053_1k,
            "bricks")
        sphere.normalTexture = GlTexture.newTexture2D( R.drawable.bricks053_1k_normal,
            "bricks-normal")

        glClearColor(0.5f, 0.5f, 0.5f, 0.5f)
        glEnable(GL_DEPTH_TEST)
    }

    override fun onDrawFrame(arg0: GL10?) {
        glClear(GL_COLOR_BUFFER_BIT or GL_DEPTH_BUFFER_BIT)
        sphere.transforms!!.rotateAffineXYZ(rotX, rotY, 0f)
        matrices.applyModel(sphere.transforms!!)
        prog.use()
        prog.uniform(sphere, matrices, light)

        // glEnable(GL_CULL_FACE);
        sphere.model!!.draw()
    }
}
